src/azul/service/repository_controller.py

```diff
@@ -224,3 +243,42 @@ def download_file(self,
             }
         else:
             assert False
+
+    def list_sources(self,
+                     catalog: CatalogName,
+                     request: AzulRequest,
+                     ) -> JSONs:
+        authentication = request.authentication
+        plugin = self.repository_plugin(catalog)
+
+        def list_sources() -> Iterable[SourceRef]:
+            try:
+                return plugin.list_sources(authentication)
+            except PermissionError:
+                raise UnauthorizedError
+
+        if authentication is None:
+            # FIXME Determine public snapshots
+            #       https://github.com/DataBiosphere/azul/issues/2978
+            sources = list_sources()
+        else:
+            plugin_name = config.plugin_name(catalog, 'repository')
+            authentication_id = f'{plugin_name}:{authentication.identity()}'
+            try:
+                source_jsons = self._source_cache_service.get(authentication_id)
+            except CacheMiss:
+                sources = list_sources()
+                source_jsons = [source.to_json() for source in sources]
+                self._source_cache_service.put(authentication_id, source_jsons)
+            else:
+                _, source_ref_cls = plugin.source_classes()
```

Comment:
That's what I mean by bothering. See if you can hide this detail in `from_json`. In OO, there is no requirement that `type(cls.from_json()) is cls` only that `isinstance(cls.from_json(), cls)`.

Reply:
I don't think that's possible because `SourceRef` cannot know from the JSON alone which subclass to instantiate (and it seems bad/circular for the superclass to even be aware of all of its subclasses). At present, I'm convinced this conversion has to involve the plugin, but I think I can minimize the extent to which the plugin leaks these details.

Reply:
> I don't think that's possible because SourceRef cannot know from the JSON alone which subclass to instantiate

Generally speaking it can, it would have to be encoded in the JSON.

> and it seems bad/circular for the superclass to even be aware of all of its subclasses

That's a myth. It's only bad if that awareness introduces a brittle coupling. Python has a mechanism for dealing with subclasses generically (`__init_subclass__`) and the parametric polymorphism employed by the SourceRef hierarchy is another way of dealing with this.

I didn't object to the involvement of the plugin, I objected to client code having to deal with an complicated implementation detail that should have be hidden inside SourceRef or the plugin.